internal/auth/oidc/repository_auth_method_update.go

```diff
@@ -17,32 +31,561 @@ import "context"
+// If the AuthMethod contains a DiscoveryUrl for an OIDC provider, ValidateAuthMethod
 // retrieves the OpenID Configuration document. The values in the AuthMethod
 // (and associated data) are validated with the retrieved document. The issuer and
 // id token signing algorithm in the configuration are validated with the
-// retrieved document. TestAuthMethod also verifies the authorization, token,
+// retrieved document. ValidateAuthMethod also verifies the authorization, token,
 // and user_info endpoints by connecting to each and uses any certificates in the
 // configuration as trust anchors to confirm connectivity.
 //
 // Options supported are: WithPublicId, WithAuthMethod
-func (r *Repository) TestAuthMethod(ctx context.Context, opt ...Option) error {
-	panic("to-do")
+func (r *Repository) ValidateAuthMethod(ctx context.Context, opt ...Option) error {
+	const op = "oidc.(Repository).TestAuthMethod"
+	opts := getOpts(opt...)
+	var am *AuthMethod
+	switch {
+	case opts.withPublicId != "":
+		var err error
+		am, err = r.lookupAuthMethod(ctx, opts.withPublicId)
+		if err != nil {
+			return errors.Wrap(err, op)
+		}
+	case opts.withAuthMethod != nil:
+		am = opts.withAuthMethod
+	default:
+		return errors.New(errors.InvalidParameter, op, "neither WithPublicId(...) nor WithAuthMethod(...) options were provided")
+	}
+
+	if err := am.isComplete(); err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	// FYI: once converted to an oidc.Provider, any certs configured will be used as trust anchors for all HTTP requests
+	provider, err := convertToProvider(ctx, am)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	info, err := provider.DiscoveryInfo(ctx)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	var result *multierror.Error
+	if info.Issuer != am.DiscoveryUrl {
+		result = multierror.Append(result, errors.New(errors.InvalidParameter, op,
+			fmt.Sprintf("auth method issuer doesn't match discovery issuer: expected %s and got %s", am.DiscoveryUrl, info.Issuer)))
+	}
+	for _, a := range am.SigningAlgs {
+		if !strutil.StrListContains(info.IdTokenSigningAlgsSupported, a) {
+			result = multierror.Append(result, errors.New(errors.InvalidParameter, op,
+				fmt.Sprintf("auth method signing alg is not in discovered supported algs: expected %s and got %s", a, info.IdTokenSigningAlgsSupported)))
+		}
+	}
+	providerClient, err := provider.HTTPClient()
+	if err != nil {
+		result = multierror.Append(result, errors.New(errors.Unknown, op, "unable to get oidc http client", errors.WithWrap(err)))
+		return result.ErrorOrNil()
+	}
+
+	// we need to prevent redirects during these tests... we don't want to have
+	// redirects going to the controller's callback (aka the configured provider's callback)
+	providerClient.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+		return http.ErrUseLastResponse
+	}
+
+	// test JWKs URL
+	statusCode, err := pingEndpoint(ctx, providerClient, "JWKs", "GET", info.JWKSURL)
```

Comment:
Any reason it's `JWKSURL` instead of `JwksUrl`? I'm guessing it's for legacy compatibility with Vault -- although that can easily be handled (and likely already is) with JSON tags -- but just checking.

Reply:
the `info` struct is imported from CAP and it the initialisms standard from  https://github.com/golang/go/wiki/CodeReviewComments#initialisms